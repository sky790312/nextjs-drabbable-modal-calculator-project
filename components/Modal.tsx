import React, { ReactNode, useState, useEffect, useRef } from 'react'
import { createPortal } from 'react-dom'
import styled from 'styled-components/'

const isClient = () => typeof window !== 'undefined'

interface Props extends StyledModalProps {
  children: ReactNode
  isShow: boolean
  closeModal: () => void
}

type StyledModalProps = {
  view: 'pc' | 'mobile'
}

const defaultStyle: StyledModalProps = {
  view: 'pc',
}

export const Modal: React.FC<Props> = ({
  children,
  isShow = false,
  view = defaultStyle.view,
  closeModal,
}) => {
  const bodyOverflow = useRef<string>(
    (isClient() && window.getComputedStyle(document.body).overflow) ?? '',
  )

  useEffect(() => {
    document.body.style.overflow = isShow ? 'hidden' : bodyOverflow.current
  }, [isShow])

  const handleCloseModal = (event: {target: EventTarget, currentTarget: EventTarget}) => {
    if (event.target === event.currentTarget) {
      closeModal()
    }
  }

  const modalContentRef = useRef(null)
  const [mouseDown, setMouseDown] = useState<boolean>(false)

  useEffect(() => {
    const handleMouseUp = () => setMouseDown(false)

    window.addEventListener('mouseup', handleMouseUp)

    return () => {
      window.addEventListener('mouseup', handleMouseUp)
    }
  }, [])

  useEffect(() => {
    const handleDrag = (movementX: number, movementY: number) => {
      const draggableContent = modalContentRef.current
      if (!draggableContent) {
        return
      }

      const { x, y } = draggableContent.getBoundingClientRect()

      draggableContent.style.left = `${x + movementX}px`
      draggableContent.style.top = `${y + movementY}px`
    }

    const handleMouseMove = (e: MouseEvent) =>
      handleDrag(e.movementX, e.movementY)

    if (mouseDown) {
      window.addEventListener('mousemove', handleMouseMove)
    }

    return () => {
      window.removeEventListener('mousemove', handleMouseMove)
    }
  }, [mouseDown])

  const handleMouseDown = () => setMouseDown(true)

  const modal =
    isClient() &&
    createPortal(
      <StyledModal onClick={handleCloseModal}>
        <StyledModalInner
          view={view}
          ref={modalContentRef}
          onMouseDown={handleMouseDown}
        >
          <CloseBtn onClick={handleCloseModal}>x</CloseBtn>
          <div>{children}</div>
        </StyledModalInner>
      </StyledModal>,
      document.body,
    )

  return <>{isShow && modal}</>
}

const StyledModal = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: ${({ theme }) => theme.colors.lightGray};
  z-index: 99;
`

const StyledModalInner = styled.div<StyledModalProps>`
  position: absolute;
  background-color: ${({ theme }) => theme.colors.white};
  border-radius: 8px;

  ${props => {
    switch (props.view) {
      case 'pc':
        return `
          min-height: 200px;
          cursor: grab;
        `
      case 'mobile':
        return `
          bottom: 0;
          width: 100%;
          height: 50%;
          overflow: auto;
        `
    }
  }}
`

const CloseBtn = styled.div`
  position: absolute;
  right: 10px;
  top: 4px;
  font-size: 20px;
  color: ${({ theme }) => theme.colors.color};
`
