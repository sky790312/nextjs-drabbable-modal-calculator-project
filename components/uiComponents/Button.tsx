import React, { ReactNode } from 'react'
import styled from 'styled-components/'

interface Props extends StyledButtonProps {
  children: ReactNode
  onClick: () => void
}

type StyledButtonProps = {
  size?: 'md' | 'block'
  buttonStyle?: 'blue' | 'gray' | 'darkGray'
}

const defaultStyle: StyledButtonProps = {
  size: 'md',
  buttonStyle: 'gray'
}

export const Button: React.FC<Props> = React.memo(({ children, size = defaultStyle.size, buttonStyle = defaultStyle.buttonStyle, onClick }) => {
  return <StyledButton buttonStyle={buttonStyle} size={size} onClick={onClick}>{children}</StyledButton>;
});

const StyledButton = styled.button<StyledButtonProps>`
  border: none;
  outline: none;
  font-size: 24px;

  ${props => {
    switch (props.buttonStyle) {
      case 'gray':
        return `
          color: ${props.theme.colors.black};
          background-color: ${props.theme.colors.gray};
        `
      case 'blue':
        return `
          color: ${props.theme.colors.white};
          background-color: ${props.theme.colors.blue};
        `
      case 'darkGray':
        return `
          color: ${props.theme.colors.white};
          background-color: ${props.theme.colors.darkGray};
        `
    }
  }}

  ${props => {
    switch (props.size) {
      case 'md':
        return `
          width: 80px;
          height: 80px;
          border-radius: 50%;
        `
      case 'block':
        return `
          width: 100%;
          height: 80px;
          border-radius: 40px;
        `
    }
  }}
`