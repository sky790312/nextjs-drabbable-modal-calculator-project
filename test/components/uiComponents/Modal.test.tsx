import React from 'react'
import { render, withProviders } from '@/test/testUtils'
import { Modal } from '@/components/uiComponents/Modal'

describe('CalculateSection component', () => {
  const modalConfig = {
      isShow: true,
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      closeModal: () => {},
    }

  it('matches snapshot', () => {
    const { asFragment } = render(withProviders(<Modal {...modalConfig}>modal</Modal>), {})
    expect(asFragment()).toMatchSnapshot()
  })
})
