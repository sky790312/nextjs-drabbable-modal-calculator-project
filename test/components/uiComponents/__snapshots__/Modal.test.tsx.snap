// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalculateSection component matches snapshot 1`] = `<DocumentFragment />`;
